### ISSUE TITLE
Service Apply Rule has no check_command definition

### ISSUE THREAD

**Issue by Tuetensuppe** [Icinga/icingaweb2-module-director#1259]
The service apply rule is missing the check_command field. This field is only customization in service templates.

## Expected Behavior
I would expect that i can create a service template with all check_execution parameters for my environment.
Then i add an apply rule which imports this template and then add the check_command which i want to execute with this rule. Therefore i would only need a few templates for all apply rules.

## Current Behavior
Currently i need to add a template for each service apply rule i want to create. As there is no option to set the check_command in the apply rule a separate template for each check_command is necessary. 
This is causing a huge amount of very similar (only different check_command) templates in my configuration.

**Comment by tigerlee** [comment]
+1, hope fix this issue or give us a workaround.

**Comment by mengine23** [comment]
+1

**Comment by dbuelow** [comment]
+1

**Comment by tigerlee** [comment]
hey guys, I found the way to fix this issue. We may use `Apply Rule` in a wrong way.

`Apply Rule` NEEDS work with service templates. You SHOULD create a `Service Template` first, then create `Apply Rule` to specify the rule and the relate `Service Template`.

**Comment by dbuelow** [comment]
Maybe i'm missing something..already created a service_template:

```
template Service "generic-service" {
    max_check_attempts = "3"
    check_period = "24x7"
    check_interval = 5m
    retry_interval = 30s
    enable_notifications = true
    enable_active_checks = true
    enable_passive_checks = false
    enable_event_handler = false
    enable_perfdata = true
    volatile = false
}
```
It is possible to reference to it via imports* but there isn't any option to specify the command.

**Comment by Thomas-Gelf** [comment]
The hard distinction between templates and objects/apply rules is intentional, and there are currently no plans to change this. Someone with the permission to choose a custom check command is able to circumvent all your restrictions, therefore this task is for admins only. In the GUI, admin-only properties are mostly accessible only via templates.

Sorry for the inconvenience, but at the time being I'm pretty strict here. @tigerlee is right, and it should work fine that way. And it does in lots of environments ;-) @dbuelow: your template is missing a command, you'll not be provided the possibility to set/switch command directly on the Apply Rule.

**Comment by dbuelow** [comment]
I understood the regular (by design) way as documented [here](https://blog.netways.de/2017/07/14/apply-service-for-im-icinga-director/). I also have other templates like:

```
template Service "mailq-service-template" {
    import "generic-service"

    check_command = "mailq"
    command_endpoint = host_name
}
```
This works fine so far. But as posted by @Tuetensuppe it would be great if there is any option to specify the check_command later inside the apply rule and to use one template for a lot of "more simple" checks. 

Hopefully i understood it...:) Thank you very much for your response.

**Comment by offsides** [comment]
Given that the default Icinga2 config that ships with the software places check commands directly inside apply rules, it's frustrating and confusing that the specific examples provided by the default Icinga2 config are explicitly not possible using Director.  This also essentially requires a service template for each and every command you want to use, even if all of the settings other than the check_command are already being imported from another template.

I can understand the concern about bypassing permissions, but the reality is that at least currently you almost require admin permissions to do anything in Director, making the distinction nearly moot.  I could understand the separation more if there were actual granularity in the permissions, but as of right now you cannot access the apply rules without admin permissions anyway...

**Comment by MarcusCaepio** [comment]
> it's frustrating and confusing that the specific examples provided by the default Icinga2 config are explicitly not possible using Director.

 
I am totally with you. I checked the Director back when the first versions were released. Back in these days, I did not used it further more, because I already had an intact monitoring with +500 hosts based on +10000k line config, which could not be imported into the director. 
Now I am at a point, where I am planning a completely new monitoring environment and thought, I test the director again, to maybe use it completely for the new environment and don't write a single config on CLI anymore. I am right now sitting here, running the Icinga2 in Docker and want to "create" a test enviroment in a docker network. It is already failing by creating a apply rule for ping, which brought me here. So I am right now, after a few clicks, at a point, where I have no desire anymore. Instead of just typing
```
apply Service "Ping" {
import "generic-service"
check_command = "ping4"
assign where host.address
}
```
I now have to create a Template for this service first and then create the assign rule, what, imho, makes no sense at all. In templates are variables, which are there to be used for XXX Services. That's why there are templates. But a template for ping, based on my generic-host template just to assign an apply rule would end in my case in way too much templates. I am really curios what and how happens, when I want to create an service with an "for" loop like tcp,http and so on
@Thomas-Gelf What's the problem to add "check_command" field?

**Comment by Thomas-Gelf** [comment]
@MarcusCaepio: guess this question has already been answered above, I'll close this issue. If you want to write plain-text configuration please feel free to do so. If you want to build a highly automated system driven by multiple data sources with the possibility to delegate day-to-day changes to your coworkers either write your own abstraction layer or give the Director a try. Most people do not want other teams to be able break things, that's why there are some artificial restrictions.

From my personal experience I can tell you that it runs fine even with a few hundred thousand checks and probably more. Technically, performance-wise and from an organizational point of view. As of this writing a freshly installed Director is "empty", you need to invest some time before he es able to save your's. We are working on providing more tooling allowing people to export and share parts of their building blocks, that will improve things a lot. You'll be able to export whole Service Sets with all related templates, Field Definitions and Commands. Or your Import/Sync definitions for Active Directory, AWS, vSphere or just a bunch of YAML files. But as of today you have to click them on your own. Usually investing that time doesn't hurt, learning by your own errors helps you to better understand the tool.

If creating a Ping template is too much work then you should better stay away from the Director, it's the wrong tool for you.

Cheers,
Thomas

**Comment by MarcusCaepio** [comment]
Well, I did not wanted to offend you personally and I am very sorry, that this issue is closed with the won't fix label as it seems I am not the only one missing this we are just splitting here between how it is done in the director and how it is described in the official doc: https://www.icinga.com/docs/icinga2/latest/doc/17-language-reference/#apply

**Comment by Thomas-Gelf** [comment]
Don't worry, I do not feel offended. I know that saying "no" and rejecting an issue feels harsh, but it wan't meant that way.

For a better understanding, let me add a few words: It's technically close to impossible to write a configuration front-end for a DSL, especially not when it should work at scale. Director and the Icinga DSL follow different goals, and for the Director the DSL is more a problem than being helpful. Many of the DSL examples Icinga 2 newcomers are being taught are there to either show how cool the DSL is or to work around shortcomings of a text-based configuration. Of course it's VERY cool and extremely powerful, and if that is what you love than you should stick with writing code and stay away from the Director. Or eventually do both, automate adding/removing hosts with the Director and define services in the DSL.

I know that people hate it being forced to first create their own templates before being able to see a result. Especially those managing only a very few hosts. The larger your environment, the more checks you deploy, the more people work together on your configuration - the more you'll learn to love some of the artificial restrictions being in place.

I already relaxed some of the restrictions we had in the past. But always not before fully understanding their impact, especially performance-wise in large environments with ACLs/restrictions in place. Being able to choose a custom Check Command is a powerful weapon, it would allow bad guys to take over the whole monitoring system. Currently Templates and Commands are accessible to Director Admins only, "normal" users are allowed to manage only "normal" objects.

Currently both are to be found in the very same database table, so adding the Check Command field to the Apply Rule Form would be very simple. But we'll probably go one more step to distinct Templates from Objects, I want to put them in separate tables. Director is very fast, but this would make largish environments even faster. As of today I'm unsure whether or not we're going to introduce more granular restrictions, currently large parts of them are route-based. I'm strictly against relaxing some of those restrictions too early just to safe others a few minutes, as this has the potential to cost me days if not weeks to figure out migrations paths when it goes to implementing what I have in mind right now. Eventually one day I might even allow what you're asking for, who knows. But please do not expect this to happen in the near future.

I hope I have been able to give you a partial understanding for my reasoning behind this decision. Director is not and will never be a 1:1 copy of your plain-text configuration. If I recall it correctly that's one of the first statements in the "Getting Started" documentation. It will however allow you to do many other different tricks with the potential to safe you a lot of time. Imagine your Ping being part of a Set of Services, this Set being applied via Apply Rule to all your hosts, 200 of those hosts have different thresholds - and 30 do not want to have that check at all. Lot of fun in your plain-text configuration I guess. Sure, it's possible - but no fun at all if you're being forced to remodel your configuration again and again when such requirements pop up over time.

With the Director overriding Thresholds has been a thing for a long time, and with v1.5.0 there will be a nice new feature allowing your "end users" to also just "Delete/Blacklist" services they got via apply rule. But this is just one simple example I recall right now, as I pushed it [yesterday](https://github.com/Icinga/icingaweb2-module-director/issues/907). Sure, you can of course do all of this also in a manually crafted configuration. But you'll loose a lot of transparency. Clicking a host, seeing why it got a specific service or that it WOULD get such service but it got blacklisted? That's extremely helpful, especially in growing environments.

Regards,
Thomas

**Comment by MarcusCaepio** [comment]
I understand your point and I did not know, that director and DSL shall not go the same way. Shame on me I did not read this in the director. If the hybrid way of configuring via DSL and import hosts automated via director and SQL CMDB works, I think this is the way I want to go. Already testing this.... :D Maybe the time will come, were I have the need to make more over the Director (the black list part looks VERY interesting). Thanks a lot Tom :)
